File: sdk/csharp/src/AnyVali/ValidationError.cs
namespace AnyVali;

/// <summary>
/// Exception thrown by Parse() when validation fails.
/// </summary>
public sealed class ValidationError : Exception
{
    public IReadOnlyList<ValidationIssue> Issues { get; }

    public ValidationError(IReadOnlyList<ValidationIssue> issues)
        : base(FormatMessage(issues))
    {
        Issues = issues;
    }

    private static string FormatMessage(IReadOnlyList<ValidationIssue> issues)
    {
        return string.Join("\n", issues.Select(i =>
        {
            var pathStr = i.Path.Count > 0
                ? string.Join(".", i.Path) + ": "
                : "";
            return $"[{i.Code}] {pathStr}{i.Message}";
        }));
    }
}

File: sdk/ruby/lib/anyvali/validation_error.rb
# frozen_string_literal: true

module AnyVali
  class ValidationError < StandardError
    attr_reader :issues

    def initialize(issues)
      @issues = issues
      messages = issues.map { |i| "#{i.code} at #{i.path.inspect}: expected #{i.expected}, received #{i.received}" }
      super("Validation failed: #{messages.join('; ')}")
    end
  end
end
